rust/src/operations/mod.rs

```diff
@@ -0,0 +1,285 @@
+    ArrowError {
+        /// Raw internal StorageError
+        #[from]
+        source: ArrowError,
+    },
+
+    /// Error returned for errors internal to Datafusion
+    #[error("Error handling arrow data")]
+    DataFusionError {
+        /// Raw internal DataFusionError
+        #[from]
+        source: DataFusionError,
+    },
+}
+
+fn to_datafusion_err(e: impl std::error::Error) -> DataFusionError {
+    DataFusionError::Plan(e.to_string())
+}
+
+/// High level interface for executing commands against a DeltaTable
+pub struct DeltaCommands {
+    table: DeltaTable,
+}
+
+impl DeltaCommands {
+    /// load table from uri
+    pub async fn try_from_uri<T>(uri: T) -> DeltaCommandResult<Self>
+    where
+        T: Into<String>,
+    {
+        let table = get_table_from_uri_without_update(uri.into())?;
+        Ok(Self { table })
+    }
+
+    /// Get a reference to the underlying table
+    pub fn table(&self) -> &DeltaTable {
+        &self.table
+    }
+
+    async fn execute(
+        &mut self,
+        operation: DeltaOperation,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> DeltaCommandResult<()> {
+        let transaction = Arc::new(DeltaTransactionPlan::new(
+            self.table.table_uri.clone(),
+            plan,
+            operation,
+            None,
+        ));
+
+        let session_ctx = SessionContext::new();
+        let task_ctx = session_ctx.task_ctx();
+        let _ = collect(transaction, task_ctx).await?;
+        self.table.update().await?;
+
+        Ok(())
+    }
+
+    /// Create a new Delta table
+    pub async fn create(
+        &mut self,
+        metadata: DeltaTableMetaData,
+        mode: Option<SaveMode>,
+    ) -> DeltaCommandResult<()> {
+        let operation = DeltaOperation::Create {
+            mode: mode.clone().unwrap_or(SaveMode::Ignore),
+            metadata: metadata.clone(),
+            location: self.table.table_uri.clone(),
+            // TODO get the protocol from somewhere central
+            protocol: Protocol {
+                min_reader_version: 1,
+                min_writer_version: 2,
+            },
+        };
+        let plan = Arc::new(CreateCommand::try_new(
+            &self.table.table_uri,
+            operation.clone(),
+        )?);
+
+        self.execute(operation, plan).await
+    }
+
+    /// Write data to Delta table
+    pub async fn write(
+        &mut self,
+        data: Vec<RecordBatch>,
+        mode: Option<SaveMode>,
+        partition_columns: Option<Vec<String>>,
+    ) -> DeltaCommandResult<()> {
+        if data.is_empty() {
+            return Ok(());
+        }
+        let schema = data[0].schema();
+        let current_part = match self.table.update().await {
+            Ok(_) => {
+                let metadata = self.table.get_metadata()?;
+                if let Some(cols) = partition_columns {
+                    if cols != metadata.partition_columns {
+                        todo!("Schema updates not yet implemented")
```

Comment:
What do you think of returning a `DeltaCommandError::UnsupportedCommand` here instead?